# Hola! 😀

Soy **Jaime Paz** y mis amigos me suelen llamar James! – *"Tranquilo, no juego en el Real Madrid"* ⚽😅. Sí, como el futbolista James Rodríguez, pero en versión científico de datos… y mejor no me digas como a él. 😉

Como tu revisor en TripleTen, estoy aquí para ayudarte a pulir tu código y tu forma de trabajar con datos. Si algo necesita un ajuste, no te preocupes: la idea es que cada comentario te acerque más a cómo se trabaja en un entorno profesional y que tu proyecto brille con todo su potencial.

Cada vez que encuentre un detalle importante en tu notebook, te lo señalaré para que puedas corregirlo y seguir creciendo. Si en algún punto no logras resolver algo, también estoy para guiarte en próximos intentos de revisión. 🔁

Es muy importante que, cuando veas mis comentarios en el notebook, **no los muevas, no los modifiques y no los borres**. Así mantenemos un historial claro de lo que ya revisamos y de tus avances. ✅

---

### Formato de Comentarios

Revisaré cuidadosamente tu notebook para asegurar que cumpla con los requisitos y te daré comentarios usando el siguiente formato:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> ✅ - ¡Excelente trabajo! Esta parte está bien implementada y contribuye de forma positiva al análisis o al proyecto. Sigue aplicando estas buenas prácticas en las siguientes secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Esta parte del código funciona, pero se puede mejorar u optimizar. Tal vez sea más claro, más eficiente o más fácil de mantener. Te señalaré ideas para que puedas reforzar esta sección.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error que es necesario corregir para aprobar esta parte. Revisa el comentario con calma, ajústalo y vuelve a intentarlo; es clave para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto. 💪
- **Áreas de mejora:** Sugerencias claras sobre lo que puedes reforzar. 💡
- **Temas adicionales para investigar:** Ideas opcionales que puedes explorar por tu cuenta para seguir creciendo.

Estos temas adicionales no son obligatorios ahora, pero pueden ayudarte a profundizar en el futuro. 📚

---

Si tienes dudas o quieres responder a un comentario específico, puedes usar este formato:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>

**¡Empecemos!** 🚀

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

    

<b>Éxito</b> ✅ - ¡Excelente trabajo! Tu proyecto es excepcional y has logrado cubrir los objetivos del mismo. Abajo he dejado mis comentarios y unas pequeñas recomendaciones para que tomes en cuenta en un futuro y te hagan ¡un máster en Data!

¡Felicidades!    

</div>

In [15]:
import pandas as pd

# Cargar el dataset
df = pd.read_csv('/datasets/users_behavior.csv')

# Mostrar primeras filas
df.head()


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [2]:
# Información general
df.info()

# Estadísticas descriptivas
df.describe()

# Verificar valores faltantes
df.isna().sum()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

# Exploración inicial de los datos

El dataset contiene 3,214 observaciones y 5 variables relacionadas con el comportamiento mensual de los usuarios de Megaline.

Las variables explicativas son:
- `calls`: número de llamadas realizadas
- `minutes`: duración total de las llamadas en minutos
- `messages`: número de mensajes de texto enviados
- `mb_used`: tráfico de internet utilizado en MB

La variable objetivo es:
- `is_ultra`: plan contratado en el mes actual (1 = Ultra, 0 = Smart)

Todas las variables son numéricas. No se detectaron valores faltantes en ninguna columna, por lo que no es necesario realizar procesos de limpieza adicionales.  
El dataset se encuentra listo para ser utilizado directamente en la etapa de modelado.


In [3]:
# Primero separamos features y target
from sklearn.model_selection import train_test_split

# Variables independientes y variable objetivo
X = df.drop('is_ultra', axis=1)
y = df['is_ultra']


In [4]:
# Usamos una división 60% entrenamiento, 20% validación, 20% prueba.
# Separar conjunto de entrenamiento (60%) y conjunto temporal (40%)
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.4, random_state=12345
)

# Separar validación (20%) y prueba (20%)
X_valid, X_test, y_valid, y_test = train_test_split(
    X_temp, y_temp, test_size=0.5, random_state=12345
)


In [5]:
print(X_train.shape)
print(X_valid.shape)
print(X_test.shape)


(1928, 4)
(643, 4)
(643, 4)


# Segmentación de los datos

Los datos se dividieron en tres conjuntos independientes:

- Conjunto de entrenamiento (60%) para entrenar los modelos.
- Conjunto de validación (20%) para ajustar hiperparámetros y comparar modelos.
- Conjunto de prueba (20%) para evaluar el rendimiento final del modelo.

Esta proporción permite entrenar el modelo con suficiente información y, al mismo tiempo, reservar datos independientes para una evaluación objetiva del desempeño.


In [6]:
#Necesitamos importar lo siguiente
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


In [7]:
#Primero Arbol de decisión
best_tree_acc = 0
best_tree_depth = None

for depth in range(1, 11):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(X_train, y_train)
    preds = model.predict(X_valid)
    acc = accuracy_score(y_valid, preds)
    
    if acc > best_tree_acc:
        best_tree_acc = acc
        best_tree_depth = depth

best_tree_acc, best_tree_depth


(0.7853810264385692, 3)

In [8]:
#Segundo Random Forest
best_forest_acc = 0
best_params = None

for n in [10, 50, 100]:
    for depth in [5, 10, None]:
        model = RandomForestClassifier(
            n_estimators=n,
            max_depth=depth,
            random_state=12345
        )
        model.fit(X_train, y_train)
        preds = model.predict(X_valid)
        acc = accuracy_score(y_valid, preds)
        
        if acc > best_forest_acc:
            best_forest_acc = acc
            best_params = (n, depth)

best_forest_acc, best_params


(0.7947122861586314, (100, 5))

In [9]:
#Tercero regresión logística
model_lr = LogisticRegression(max_iter=1000, random_state=12345)
model_lr.fit(X_train, y_train)

preds_lr = model_lr.predict(X_valid)
lr_acc = accuracy_score(y_valid, preds_lr)

lr_acc


0.7107309486780715

# Evaluación de modelos y ajuste de hiperparámetros

Se evaluaron tres modelos de clasificación: Árbol de Decisión, Random Forest y Regresión Logística.

El Árbol de Decisión mostró un desempeño aceptable en profundidades bajas, pero comenzó a sobreajustar conforme aumentó la profundidad.

El modelo Random Forest obtuvo el mejor desempeño general, logrando una mayor exactitud en el conjunto de validación gracias a la combinación de múltiples árboles, lo que reduce el sobreajuste.

La Regresión Logística presentó un rendimiento estable, aunque inferior a los modelos basados en árboles, lo cual es consistente con su naturaleza lineal.

Con base en los resultados obtenidos en el conjunto de validación, el modelo Random Forest fue seleccionado como el mejor candidato para la evaluación final.


In [10]:
# Seleccionamos el Modelo Random Forest, asi que este se evaluará
# Primero, entrenamos el modelo final con los mejores parámetros encontrados:

final_model = RandomForestClassifier(
    n_estimators=100,
    max_depth=5,
    random_state=12345
)

final_model.fit(X_train, y_train)


RandomForestClassifier(max_depth=5, random_state=12345)

In [11]:
# Segundo evaluamos en el conjunto de prueba:
from sklearn.metrics import accuracy_score

test_preds = final_model.predict(X_test)
test_acc = accuracy_score(y_test, test_preds)

test_acc


0.7900466562986003

#Evaluación del modelo en el conjunto de prueba

El modelo final seleccionado fue Random Forest con 100 árboles y una profundidad máxima de 5.  
Al evaluarlo en el conjunto de prueba, se obtuvo una exactitud de aproximadamente 0.79, superando el umbral mínimo requerido de 0.75.

Este resultado indica que el modelo generaliza adecuadamente y puede utilizarse para recomendar planes móviles a nuevos clientes.


In [12]:
# 1.- Primero, veamos cuál es la clase mayoritaria en el set de entrenamiento:
y_train.value_counts(normalize=True)


0    0.692427
1    0.307573
Name: is_ultra, dtype: float64

In [13]:
# 2.- Baseline: siempre predecir la clase mayoritaria y medir accuracy en test:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

dummy = DummyClassifier(strategy='most_frequent', random_state=12345)
dummy.fit(X_train, y_train)

dummy_preds = dummy.predict(X_test)
dummy_acc = accuracy_score(y_test, dummy_preds)

dummy_acc


0.6842923794712286

In [14]:
# 3.- Comparación directa (modelo vs baseline):
print("Accuracy Random Forest (test):", test_acc)
print("Accuracy baseline (most_frequent):", dummy_acc)
print("Diferencia:", test_acc - dummy_acc)


Accuracy Random Forest (test): 0.7900466562986003
Accuracy baseline (most_frequent): 0.6842923794712286
Diferencia: 0.10575427682737171


# Prueba de cordura del modelo

Como prueba de cordura, se entrenó un modelo baseline utilizando `DummyClassifier`, el cual siempre predice la clase mayoritaria (plan Smart).

La clase mayoritaria representa aproximadamente el 69% de los usuarios, y el modelo baseline obtuvo una exactitud de 0.68 en el conjunto de prueba.

El modelo Random Forest alcanzó una exactitud de 0.79, superando al baseline por más de 10 puntos porcentuales.  
Esto confirma que el modelo aprende patrones significativos a partir de los datos y no se limita a predecir la clase más frecuente.


# Conclusiones y recomendación final

En este proyecto se desarrolló un modelo de clasificación para recomendar planes móviles (Smart o Ultra) con base en el comportamiento mensual de los usuarios.  
El modelo Random Forest seleccionado alcanzó una exactitud de 0.79 en el conjunto de prueba, superando el umbral requerido y demostrando una mejora significativa frente a un modelo baseline.

Los resultados indican que el comportamiento de uso —especialmente el consumo de minutos y datos— es un factor relevante para identificar qué plan se ajusta mejor a cada usuario.  
Por ello, se recomienda a Megaline utilizar este modelo como una herramienta de apoyo para la toma de decisiones comerciales, particularmente en estrategias de recomendación personalizada y upselling.

El modelo no debe considerarse una decisión automática, sino un sistema de recomendación que complemente el criterio comercial. Para una implementación en producción, sería conveniente incorporar variables adicionales como antigüedad del cliente, historial de pagos y métricas de abandono (churn), así como evaluar métricas adicionales de desempeño.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

Hola Veronica! buenas noticias, tienes los requerimientos escenciales completados ✅. Has cubierto los ítems principales requeridos: cargaste el dataset, el notebook corre sin errores, reconoces las features, implementaste varios modelos y realizaste la división 60/20/20 con random_state. Además evaluaste en test y realizaste una prueba de cordura (baseline). 💡  

**Sugerencias rápidas**: considera usar GridSearchCV/RandomizedSearchCV o cross-validation (si quieres alcanzar un nivel intermedio / avanzado) para una búsqueda de hiperparámetros más robusta, reportar métricas adicionales (F1, AUC-ROC) y mostrar una matriz de confusión o feature importance para interpretar mejor el modelo. 🚀
</div>

### Conclusiones y recomendaciones

- Nivel general: en conjunto, este trabajo se sitúa más en un nivel intermedio: cumple los requisitos del proyecto y muestra buena práctica en la comparación y evaluación de modelos. ✅  
- Fortalezas principales:  
  - División de datos correcta (60/20/20) y uso de random_state para reproducibilidad. ✅  
  - Implementaste y comparaste varios modelos (Decision Tree, Random Forest, Logistic Regression) y seleccionaste el mejor con base en validación y prueba. ✅  
  - Realizaste una prueba de cordura con DummyClassifier y comparaste el beneficio frente al baseline. ✅  
- Áreas de oportunidad y mejora:  
  - **Sería bueno incorporar GridSearchCV o cross-validation** para una optimización de hiperparámetros más sólida y reproducible. Con esto, vas a alcanzar el nivel AVANZADO   
  - **RECOMENDADO** Añadir métricas adicionales (F1, AUC-ROC) y visualizaciones (matriz de confusión, importancias de features, distribuciones) para entender mejor errores y el comportamiento del modelo. 📊  
  - **RECOMENDADO: Profundizar en interpretación** (por qué se confunden ciertas clases, posibles feature engineering o outliers) para llevar el modelo hacia un nivel más avanzado. 🔁  
- Nota sobre criterios técnicos: los criterios obligatorios del proyecto están cumplidos. Toma en cuenta que, para subir de intermedio a avanzado, conviene reforzar validación cruzada, más métricas y análisis interpretativo. 💪🚀